pkg/generate/cspc.go

```diff
@@ -0,0 +1,215 @@
+	var spec []cstorv1.PoolSpec
+	if poolType == string(cstorv1.PoolStriped) {
+		// always a single RAID-group with nDevices patched together, cannot disk replace,
+		// no redundancy in a pool, redundancy possible across pool instances
+
+		// for each eligible set of BDs from each eligible node, take nDevices number of BDs
+		for _, node := range nodes {
+			bds := bd[node]
+			var raid cstorv1.RaidGroup
+			for d := 0; d < nDevices; d++ {
+				raid = cstorv1.RaidGroup{
+					CStorPoolInstanceBlockDevices: []cstorv1.CStorPoolInstanceBlockDevice{{BlockDeviceName: bds[d].Name}}}
+			}
+			spec = append(spec, cstorv1.PoolSpec{
+				NodeSelector:   map[string]string{"kubernetes.io/hostname": node},
+				DataRaidGroups: []cstorv1.RaidGroup{raid},
+				PoolConfig: cstorv1.PoolConfig{
+					DataRaidGroupType: string(cstorv1.PoolStriped),
+				},
+			})
+		}
+		return &spec, nil
+	} else if poolType == string(cstorv1.PoolMirrored) {
+		if nDevices%2 != 0 {
+			return nil, fmt.Errorf("mirrored pool requires multiples of two block device")
+		}
+		for node, bds := range bd {
+			var raids []cstorv1.CStorPoolInstanceBlockDevice
+			// add all BDs to a CSPCs CSPI spec
+			for d := 0; d < nDevices; d++ {
+				raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[d].Name})
+			}
+			// add the CSPI BD spec inside CSPC to a PoolSpec
+			spec = append(spec, cstorv1.PoolSpec{
+				NodeSelector:   map[string]string{"kubernetes.io/hostname": node},
+				DataRaidGroups: []cstorv1.RaidGroup{{CStorPoolInstanceBlockDevices: raids}},
+				PoolConfig: cstorv1.PoolConfig{
+					DataRaidGroupType: string(cstorv1.PoolMirrored),
+				},
+			})
+		}
+		return &spec, nil
+		// 2ⁿ devices per RaidGroup, (confirm) not more than 2 devices per RaidGroup
+		// DOUBT: Should this throw an error if nDevices isn't 2ⁿ?
+	} else if poolType == string(cstorv1.PoolRaidz) {
+		return nil, fmt.Errorf("%s is not supported yet", poolType)
+		// 2ⁿ+1 devices per RaidGroup
+	} else if poolType == string(cstorv1.PoolRaidz2) {
+		return nil, fmt.Errorf("%s is not supported yet", poolType)
+		// 2ⁿ+2 devices per RaidGroup
+	}
+	return nil, fmt.Errorf("unknown pool-type")
+}
+
+// filterCStorCompatible takes a list of BDs and gives out a list of BDs which can be used to provision a pool
+func filterCStorCompatible(bds *v1alpha1.BlockDeviceList) (*v1alpha1.BlockDeviceList, error) {
+	// TODO: Optionally reject sparse-disks depending on configs
+	var final []v1alpha1.BlockDevice
+	for _, bd := range bds.Items {
+		// an eligible blockdevice is in active+unclaimed state and lacks a file-system
+		if bd.Status.State == v1alpha1.BlockDeviceActive &&
+			bd.Status.ClaimState == v1alpha1.BlockDeviceUnclaimed &&
+			bd.Spec.FileSystem.Type == "" &&
+			// BD's capacity >=64 MiB
+			bd.Spec.Capacity.Storage >= 67110000 {
+			final = append(final, bd)
+		}
+	}
```

Comment:
Yeah, Good to reject sparse-based blockdevices(Good Point :+1: ),

We can replace the above snippet with something like below(it can be improved further):
```go
type poolBDFilter func(v1alpha1.BlockDevice) bool
var (
poolEligibleBDFilters := []poolBDFilter{
  isBlockDeviceActive,
  isBlockDeviceUnclaimed,
  isBlockDeviceUnFormatted,
  isNonSparseBlockDevice,
  isBlockDeviceHasMinCap(cap uint64) func(v1alpha1.BlockDevice) bool {
      return func(bd v1alpha1.BlockDevice) bool {
           return bd.Spec.Capacity.Storage >= minCapacity
      }
  }
}
)

func isBlockDeviceEligibleForPoolCreation(bd v1alpha1.BlockDevice, filters []poolBDFilter) bool {
             for i:=0; i<len(filters); i++ {
                   if !filters[i](bd) {
                       return false
                  }
             }
             return true
}

if isBlockDeviceEligibleForPoolCreation(bd, poolEligibleBDFilters) {
   final = append(final, bd)
}
```